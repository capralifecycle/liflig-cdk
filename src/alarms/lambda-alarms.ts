import * as constructs from "constructs"
import * as cdk from "aws-cdk-lib"
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch"

export interface LambdaAlarmsProps {
  actions: cloudwatch.IAlarmAction[]
  lambdaFunctionName: string
}

export class LambdaAlarms extends constructs.Construct {
  private readonly actions: cloudwatch.IAlarmAction[]
  private readonly lambdaFunctionName: string

  constructor(
    scope: constructs.Construct,
    id: string,
    props: LambdaAlarmsProps,
  ) {
    super(scope, id)

    this.actions = props.actions
    this.lambdaFunctionName = props.lambdaFunctionName
  }

  /**
   * Sets up a CloudWatch Alarm that triggers if the Lambda fails invocations.
   * This usually happens from uncaught exceptions in the lambda.
   */
  addInvocationErrorAlarm(
    /**
     * Configuration for an alarm.
     */
    props?: {
      /**
       * Add extra information to the alarm description, like Runbook URL or steps to triage.
       */
      appendToAlarmDescription?: string
    },
  ): void {
    const alarm = new cloudwatch.Metric({
      metricName: "Errors",
      namespace: "AWS/Lambda",
      statistic: "Sum",
      period: cdk.Duration.seconds(60), // Standard resolution metric has a minimum of 60s period
      dimensionsMap: {
        FunctionName: this.lambdaFunctionName,
      },
    }).createAlarm(this, "FailedInvocationAlarm", {
      alarmDescription: `Invocation for '${this.lambdaFunctionName}' failed. ${props?.appendToAlarmDescription ?? ""}`,
      comparisonOperator:
        cloudwatch.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      evaluationPeriods: 1,
      threshold: 1,
      treatMissingData: cloudwatch.TreatMissingData.IGNORE,
    })

    this.actions.forEach((action) => {
      alarm.addAlarmAction(action)
      alarm.addOkAction(action)
    })
  }
}
