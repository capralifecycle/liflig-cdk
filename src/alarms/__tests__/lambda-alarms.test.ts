import "@aws-cdk/assert/jest"
import { App, Stack } from "aws-cdk-lib"
import * as cloudwatchActions from "aws-cdk-lib/aws-cloudwatch-actions"
import * as lambda from "aws-cdk-lib/aws-lambda"
import * as sns from "aws-cdk-lib/aws-sns"
import "jest-cdk-snapshot"
import { LambdaAlarms } from "../lambda-alarms"

test("create alarms", () => {
  const app = new App()
  const supportStack = new Stack(app, "SupportStack")
  const stack = new Stack(app, "Stack")

  const topic = new sns.Topic(supportStack, "Topic")
  const action = new cloudwatchActions.SnsAction(topic)

  const lambdaFunction = new lambda.Function(stack, "TestedLambda", {
    runtime: lambda.Runtime.NODEJS_LATEST,
    handler: "index.handler",
    code: lambda.Code.fromInline("exports.handler = async () => {}"),
    functionName: "lambda-function-name",
  })

  const alarms = new LambdaAlarms(stack, "LambdaAlarms", {
    action: action,
    lambdaFunction: lambdaFunction,
  })

  alarms.addInvocationErrorAlarm({
    appendToAlarmDescription: "Runbook at https://liflig.no",
  })

  expect(stack).toMatchCdkSnapshot()
})
