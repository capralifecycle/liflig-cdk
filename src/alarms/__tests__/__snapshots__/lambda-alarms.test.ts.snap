// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`create alarms 1`] = `
Object {
  "Resources": Object {
    "LambdaAlarmsFailedInvocationAlarmF0FCB95D": Object {
      "Properties": Object {
        "AlarmActions": Array [
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefTopicBFC7AF6ECB4A357A",
          },
        ],
        "AlarmDescription": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Invocation for '",
              Object {
                "Ref": "TestedLambdaAAAB7B55",
              },
              "' failed. Runbook at https://liflig.no",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "FunctionName",
            "Value": Object {
              "Ref": "TestedLambdaAAAB7B55",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "Errors",
        "Namespace": "AWS/Lambda",
        "OKActions": Array [
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefTopicBFC7AF6ECB4A357A",
          },
        ],
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestedLambdaAAAB7B55": Object {
      "DependsOn": Array [
        "TestedLambdaServiceRole1EFFE97F",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "exports.handler = async () => {}",
        },
        "FunctionName": "lambda-function-name",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestedLambdaServiceRole1EFFE97F",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestedLambdaServiceRole1EFFE97F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
